staging/src/k8s.io/apimachinery/pkg/watch/mux_test.go

```diff
@@ -246,3 +245,39 @@ func TestBroadcasterSendEventAfterShutdown(t *testing.T) {
 	assert.Equal(t, sendOnClosed, false, "ActionOrDrop should return false if broadcaster is already shutdown")
 	assert.EqualError(t, err, "broadcaster already stopped", "ActionOrDrop should report error id broadcaster is shutdown")
 }
+
+// Test this since we see usage patterns where the broadcaster and watchers are
+// stopped simultaneously leading to races.
+func TestBroadcasterShutdownRace(t *testing.T) {
+	m := NewBroadcaster(0, WaitIfChannelFull)
+	stopCh := make(chan struct{})
+
+	// Add a bunch of watchers
+	const testWatchers = 2
+	for i := 0; i < testWatchers; i++ {
+		_, err := m.Watch()
+		if err != nil {
+			t.Fatalf("Unable start event watcher: '%v' (will not retry!)", err)
+		}
+		// This is how we force the watchers to close down independently of the
+		// eventbroadcaster, see real usage pattern in startRecordingEvents()
+		go func() {
+			<-stopCh
+			t.Log("Stopping Watchers")
+			m.stopWatching(int64(i))
```

Comment:
https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable

you need to capture the `i` variable with `i := i`

Reply:
```
diff --git a/staging/src/k8s.io/apimachinery/pkg/watch/mux_test.go b/staging/src/k8s.io/apimachinery/pkg/watch/mux_test.go
index 17e5779cfcd..35fe9258bfd 100644
--- a/staging/src/k8s.io/apimachinery/pkg/watch/mux_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/watch/mux_test.go
@@ -263,11 +263,17 @@ func TestBroadcasterShutdownRace(t *testing.T) {
                // eventbroadcaster, see real usage pattern in startRecordingEvents()
                go func() {
                        <-stopCh
-                       t.Log("Stopping Watchers")
+                       t.Logf("Stopping Watchers %d %d", int64(i), i)
                        m.stopWatching(int64(i))
                }()
        }
```

see 
```
=== RUN   TestBroadcasterShutdownRace
    mux_test.go:266: Stopping Watchers 2 2
    mux_test.go:266: Stopping Watchers 2 2
--- PASS: TestBroadcasterShutdownRace (0.00s)
```

Reply:
Ahh right I forgot about the var capture, that always gets me :)